WordPress/WordPressTest/NullBlogPropertySanitizerTests.swift

```diff
@@ -0,0 +1,69 @@
+import XCTest
+import Nimble
+
+@testable import WordPress
+
+class NullBlogPropertySanitizerTests: XCTestCase {
+    private var keyValueDatabase: StubKeyValueDatabase!
+    private var nullBlogPropertySanitizer: NullBlogPropertySanitizer!
+
+    private var context: NSManagedObjectContext!
+
+    private var currentBuildVersion: String {
+        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
+    }
+
+    override func setUp() {
+        super.setUp()
+        context = TestContextManager().mainContext
+        keyValueDatabase = StubKeyValueDatabase()
+        nullBlogPropertySanitizer = NullBlogPropertySanitizer(store: keyValueDatabase, context: context)
+    }
+
+    override func tearDown() {
+        super.tearDown()
+        nullBlogPropertySanitizer = nil
+        keyValueDatabase = nil
+        context = nil
+        ContextManager.overrideSharedInstance(nil)
+    }
+
+    func testSetsTheSanitizedVersionEqualToCurrentBuildVersion() {
+        keyValueDatabase.lastSanitizationVersionNumber = "10.0"
+
+        nullBlogPropertySanitizer.sanitize()
+
+        expect(self.keyValueDatabase.lastSanitizationVersionNumber).to(equal(currentBuildVersion))
+        expect(self.keyValueDatabase.setValueForKeyInvocationCount).to(equal(2))
+    }
+
+    func testDoesntChangeVersionWhenSanitizationIsNotNeeded() {
+        // Given
+        keyValueDatabase.lastSanitizationVersionNumber = currentBuildVersion
+
+        // When
+        nullBlogPropertySanitizer.sanitize()
+
+        // Then
+        // The first invocation is in the _Given_ paragraph.
+        expect(self.keyValueDatabase.setValueForKeyInvocationCount).to(equal(1))
+    }
+}
+
+private class StubKeyValueDatabase: EphemeralKeyValueDatabase {
+    private(set) var setValueForKeyInvocationCount = 0
```

Comment:
I'm not sure why we need to override this class. Testing the number of times that the database is called seems like an implementation detail that the test should not care about: we are testing whether the version changes or not, I wouldn't really care if the implementation tried to write the existing value to the database.

The `lastSanitizationVersionNumber` is nice, but also something that could be done more generically for `KeyValueDatabase`. We already have a `bool(forKey:)` helper there, we could add a `string(forKey:)`, or even a generic one:

```swift
extension KeyValueDatabase {
    subscript<T>(key: String) -> T? {
        get {
            return object(forKey: key) as? T
        }
        set(newValue) {
            set(newValue, forKey: key)
        }
    }
}
```

Then you can just call that in the test:

```swift
expect(self.keyValueDatabase[NullBlogPropertySanitizer.lastSanitizationVersionNumber])
    .to(equal(currentBuildVersion))
```

Reply:
Thanks for the feedback!

> The lastSanitizationVersionNumber is nice, but also something that could be done more generically for KeyValueDatabase.

The intent was more about making the test readable. And also to hide the implementation detail (`lastSanitizationVersionNumber`). 

I did start with using `set(,:forKey)` and the lines got pretty long. Your `subscript` makes it a bit shorter. Thanks for that! TIL

Done in 595f5bba21e53a2238e39f06c6609befb0ec9a4d and d3d7e89c90254f21de2121c6fe0489c2134f0588.

> I'm not sure why we need to override this class. Testing the number of times that the database is called seems like an implementation detail that the test should not care about

You're right about the _implementation detail_ bit. In `testDoesntChangeVersionWhenSanitizationIsNotNeeded`, we are proving that the deletion routine in `sanitize()` is **never executed** unless the app was upgraded. Unfortunately, I couldn't find an easier way to test this. 

- We can't just check the value of `NullBlogPropertySanitizer.lastSanitizationVersionNumber` in `KeyValueDatabase` because it never changes during the test scenario. 
-  Passing a mocked `NSManagedObjectContext` and testing that the `context.perform { }` was never called seems more appropriate. But mocking that in Swift is probably a bit hairy. 
- Adding `Post` records with `NULL` `blog` values would be the best. We tried that but Core Data just fails and won't allow us to save the `Post`. 

With that said, I chose the `setValueForKeyInvocationCount` hack instead. It does bleed into the _"testing the implementation detail"_ side because we _had_ to know that the `set(,forKey:)` is only called once. I usually try to avoid doing testing like this. But considering the other options, I figured this was the less intrusive option. 😞 

I think I'll keep it for now. We are hoping for `NullBlogPropertySanitizer` to be temporary and iterating more may be too much at this point. 🙂